datafusion/expr/src/type_coercion/functions.rs

```diff
@@ -130,6 +130,31 @@ fn get_valid_types(
             _ => Ok(vec![vec![]]),
         }
     }
+    fn array_append_and_optional_index(
```

Comment:
`array_element_and_optional_index` seems a better name

Reply:
Also in [5086f31](https://github.com/apache/arrow-datafusion/pull/9233/commits/5086f3150b3533f7bc7bb3e60d8561a24b5f6797)